File: csharp/monogame/SnakeGame.cs
using System;
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;
using Microsoft.Xna.Framework.Input;
using System.Collections.Generic;

namespace Snake
{
    public class SnakeGame : Game {
        GraphicsDeviceManager graphics;
        SpriteBatch spriteBatch;
        Texture2D rectTexture;
        Random rng;

        Apple apple;
        Snake snake;

        public SnakeGame() {
            graphics = new GraphicsDeviceManager(this);

            TargetElapsedTime = TimeSpan.FromMilliseconds(1000/Constants.FPS);
        }

        protected override void Initialize() {
            graphics.PreferredBackBufferHeight = Constants.SPRITE_SIZE * Constants.SCREEN_HEIGHT;
            graphics.PreferredBackBufferWidth = Constants.SPRITE_SIZE * Constants.SCREEN_WIDTH;
            graphics.ApplyChanges();

            apple = new Apple();
            snake = new Snake();

            rng = new Random();

            GenerateApple();

            base.Initialize();
        }

        private Texture2D GenerateTexture(int width, int height) {
            var data = new Color[width*height];
            for(int i=0; i<width*height;i++)
                data[i] = Color.White;

            var texture = new Texture2D(GraphicsDevice, width, height);
            texture.SetData(data);
            return texture;
        }

        protected override void LoadContent() {
            spriteBatch = new SpriteBatch(GraphicsDevice);
            rectTexture = GenerateTexture(Constants.SPRITE_SIZE - 1, Constants.SPRITE_SIZE - 1);
        }

        private void HandleInput() {
            var state = Keyboard.GetState();
            if(snake.DX == 0) {
                if(state.IsKeyDown(Keys.Left)) {
                    snake.Face(-1, 0);
                }
                if(state.IsKeyDown(Keys.Right)) {
                    snake.Face(1, 0);
                }
            }
            if(snake.DY == 0 ) {
                if(state.IsKeyDown(Keys.Up)) {
                    snake.Face(0, -1);
                }
                if(state.IsKeyDown(Keys.Down)) {
                    snake.Face(0, 1);
                }
            }
        }

        void GenerateApple() {
            do {
                apple.Move(
                    rng.Next(Constants.SCREEN_WIDTH),
                    rng.Next(Constants.SCREEN_HEIGHT)
                );
            } while(snake.CheckCollision(apple.X, apple.Y));
        }

        protected override void Update(GameTime gameTime) {
            HandleInput();

            if (GamePad.GetState(PlayerIndex.One).Buttons.Back == ButtonState.Pressed || Keyboard.GetState().IsKeyDown(Keys.Escape))
                Exit();

            snake.Update();

            if(snake.CheckCollision(apple.X, apple.Y)) {
                snake.Eat();
                GenerateApple();
            }

            base.Update(gameTime);
        }

        protected override void Draw(GameTime gameTime) {
            GraphicsDevice.Clear(Color.Black);

            var position = new Vector2(100,100);

            spriteBatch.Begin();
            apple.Draw(spriteBatch, rectTexture);
            snake.Draw(spriteBatch, rectTexture);
            spriteBatch.End();

            base.Draw(gameTime);
        }

        static void Main() {
            using (var game = new SnakeGame())
                game.Run();
        }
    }
}

File: golang/ebiten/objects/snakeGame.go
package objects

import (
	"image/color"
	"math/rand"
	"snake/constants"

	"github.com/hajimehoshi/ebiten/v2"
)

func GenerateApple(snake Snake, apple *Apple) {
	for ok := true; ok; ok = snake.CheckCollision(apple.Position) {
		apple.Move(
			rand.Intn(constants.ScreenSize),
			rand.Intn(constants.ScreenSize),
		)
	}
}

type SnakeGame struct {
	apple Apple
	snake Snake
}

func (g *SnakeGame) Update() error {
	g.snake.HandleInput()
	g.snake.Update()

	if g.snake.CheckCollision(g.apple.Position) {
		g.snake.tail++
		GenerateApple(g.snake, &g.apple)
	}
	return nil
}

func (g *SnakeGame) Draw(screen *ebiten.Image) {
	g.apple.Draw(screen)
	g.snake.Draw(screen)
}

func (g *SnakeGame) Layout(outsideWidth, outsideHeight int) (screenWidth, screenHeight int) {
	return constants.ScreenSize * constants.SpriteSize,
		constants.ScreenSize * constants.SpriteSize
}

func CreateSnakeGame() *SnakeGame {
	position := Vector{constants.ScreenSize / 2, constants.ScreenSize / 2} // start position

	apple := Apple{
		position,
		constants.SpriteSize - 1,     // width
		constants.SpriteSize - 1,     // height
		color.RGBA{0xff, 0, 0, 0xff}} // red

	snake := Snake{
		position,                     // start position
		Vector{0, 0},                 // velocity
		constants.SpriteSize - 1,     // width
		constants.SpriteSize - 1,     // height
		color.RGBA{0, 0xff, 0, 0xff}, // green
		constants.InitialTail,
		[]Vector{position},
		[]ebiten.Key{}}

	GenerateApple(snake, &apple)

	return &SnakeGame{apple, snake}
}
